components/dada-execute/src/step.rs

```diff
@@ -0,0 +1,327 @@
+use dada_id::prelude::*;
+use dada_ir::{
+    class::Class,
+    code::{
+        bir::{self, TerminatorData, TerminatorExpr},
+        syntax,
+    },
+    error,
+    origin_table::HasOriginIn,
+    span::FileSpan,
+    word::Word,
+};
+use dada_parse::prelude::*;
+
+use crate::{
+    error::DiagnosticBuilderExt,
+    kernel::Kernel,
+    machine::{op::MachineOp, ObjectData, ProgramCounter, Tuple, ValidPermissionData, Value},
+    thunk::Thunk,
+};
+
+use self::traversal::Anchor;
+
+mod access;
+mod apply_op;
+mod call;
+mod give;
+mod lease;
+mod revoke;
+mod share;
+mod tenant;
+mod traversal;
+
+pub(crate) struct Stepper<'me> {
+    db: &'me dyn crate::Db,
+    machine: &'me mut dyn MachineOp,
+    kernel: &'me mut dyn Kernel,
+    temporaries: Vec<Value>,
+}
+
+pub(crate) enum ControlFlow {
+    Next,
+    Done(Value),
+    Await(Thunk),
+}
+
+impl<'me> Stepper<'me> {
+    pub(crate) fn new(
+        db: &'me dyn crate::Db,
+        machine: &'me mut dyn MachineOp,
+        kernel: &'me mut dyn Kernel,
+    ) -> Self {
+        Self {
+            db,
+            machine,
+            kernel,
+            temporaries: vec![],
+        }
+    }
+
+    pub(crate) fn step(&mut self) -> eyre::Result<ControlFlow> {
```

Comment:
@brson 

This is the most fine-grained interface now: basically, take one step at a time. We should discuss how this fits with the REPL.